### DISCUSSION TITLE
Question: yard plugin creation

### DISCUSSION THREAD

**Discussion by LRFalk01** [lsegal/yard#1569] [Q&A] [answered]
I've been looking for a guide or walkthrough on how to create yard plugins, and have come up empty. The docs give a high level overview on the basics, but do not really go over how to actually create a plugin or how to test them.

I'd be very thankful for any blog/video/doc links I may have missed that cover plugins in more detail.

**Comment by lsegal** [comment, marked as answer]
YARD has guides on the website that walk through the process, both for building parsers and for custom templates (your plugin can do one or ther other, or even both): https://yardoc.org/guides 

If it's not fully explained there, you can peruse `yard doc help` to see how to activate plugins (packaged RubyGems) or custom code (.rb files not part of an explicit plugin).

For example code, you can look at [existing plugins](https://rubygems.org/search?query=yard-) as an example of how to build package and publish them. You can even look at YARD's own use of custom code in [the docs/ directory of this repository](https://github.com/lsegal/yard/blob/main/docs/templates/plugin.rb).

As far as testing goes, this should be no different than testing any other Ruby component. Since plugins are effectively just added or modified YARD objects, you can also look at [YARD's own tests](https://github.com/lsegal/yard/blob/main/spec/handlers/method_handler_spec.rb) to see how we do it there. Testing is going to depend entirely on what exactly you're building, so it's hard to provide specific instruction here except that YARD should not require any custom tooling to test, though if you are generating HTML content as part of your plugin, you may want to pull in some HTML specific testing tools.

Hope that all helps!
